src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,190 @@
+        tail.prev = currentNode;
+        currentNode.next = tail;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("invalid index");
+        }
+        if (index == size || head == null) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            linkBefore(value, head);
+            size++;
+            return;
+        }
+        Node<T> target = getNode(index);
+        linkBefore(value, target);
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> target) {
+        Node<T> newNode = new Node<>(value, null, null);
+        if (target == head) {
+            newNode.next = target;
+            target.prev = newNode;
+            head = newNode;
+            return;
+        }
+        Node<T> previousNode = target.prev;
+        newNode.next = target;
+        target.prev = newNode;
+        previousNode.next = newNode;
+        newNode.prev = previousNode;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("invalid index");
+        }
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode.data;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("invalid index");
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        T oldData = currentNode.data;
+        currentNode.data = value;
+        return oldData;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode;
+        if (index == 0) {
+            currentNode = head;
+            head = head.next;
+            currentNode.next.prev = null;
+            size--;
+            return currentNode.data;
+        }
+        currentNode = getNode(index);
+        if (currentNode.next == null) {
+            currentNode.prev.next = null;
+            size--;
+            return currentNode.data;
+        }
+        unlink(currentNode);
+        size--;
+        return currentNode.data;
     }
 
     @Override
     public boolean remove(T object) {
```

Comment:
Please, try to simplify this method. Maybe you can move some logic to "unlink" method